### ISSUE TITLE
Pino object logger option gets cloned after #568

### ISSUE THREAD

**Issue by stefanosala** [fastify/fastify-cli#574]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.10.2

### Plugin version

5.7.0

### Node.js version

18.12.1

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

13.0.1

### Description

After #568 was merged, the new behavior of `deepmerge` clone the `options` object and tries to clone any instance of Pino we pass as option, modifying it into a simple object so that all the Pino settings are lost.

### Steps to Reproduce

Given the following fastify startup module:

```javascript
import pino from 'pino';

export default async function (fastify, options) {
  fastify.get('/', async (req, reply) => ({ hello: 'world' }));
}

const logger = pino({
  transport: {
    target: 'pino-pretty',
  },
});

export const options = {
  logger,
};
```

When using fastify-cli like this:

    node_modules/.bin/fastify start --options a.js

I get this output:

    {"level":30,"time":1670931640310,"pid":68940,"hostname":"Stefanos-MacBook-Air.local","msg":"Server listening at http://[::1]:3000"}
    {"level":30,"time":1670931640311,"pid":68940,"hostname":"Stefanos-MacBook-Air.local","msg":"Server listening at http://127.0.0.1:3000"}

### Expected Behavior

The expected output would be something like this instead:

    [12:41:03.421] INFO (69034): Server listening at http://[::1]:3000
    [12:41:03.422] INFO (69034): Server listening at http://127.0.0.1:3000

I would tend to think that `deepmerge` should be modified in order to clone by reference any type of object that is not a plain JS object, but I'd appreciate any feedback in order to try and provide a patch for this.

**Comment by Eomm** [comment]
Thanks for reporting

Let me check what I broke

**Comment by jordanebelanger** [comment]
> Thanks for reporting
> 
> Let me check what I broke

Would be great if this can get some attention, this was really hard to debug and a huge head scratcher while I tried to figure out why my logger was no longer working as expected in production.

**Comment by Tom-Brouwer** [comment]
> Thanks for reporting
> 
> Let me check what I broke

Great, thanks for your response.

Like @jordanebelanger I have a custom logger defined in my options in app.js, and no logs were being generated anymore after an update of fastify-cli.

As mentioned by the TS, the issue is in version 5.6.1 and above. On 5.6.0 it is still working. The workaround for now is to stick to 5.6.0.

Please let me know if there's any more information you need.

**Comment by stefanosala** [comment]
@Eomm I've gone a bit into the rabbit hole and I can't seem to find an elegant solution to this issue.

`deepmerge` will try to clone the `EventEmitter` from Pino because we try to merge a `logger` option on [this line](https://github.com/fastify/fastify-cli/pull/568/files#diff-a32491fa91f5482043b89399c88b89b58e660137d926f75dab64e9c56cf454a0R138) and that seems to be "by design", so I'm not really sure how to help here, I'm sorry!

**Comment by Eomm** [comment]
I think we should explore a bit the `cloneProtoObject` parameter: https://github.com/fastify/fastify-cli/pull/568/files#diff-a32491fa91f5482043b89399c88b89b58e660137d926f75dab64e9c56cf454a0R9

If we check the obj prototype, we should be able to understand if we must clone it or not

**Comment by stefanosala** [comment]
Thanks, what type of objects do you think we should clone?
My guess would be to clone only plain objects (`Object.getPrototypeOf(value) === Object.prototype`)?

**Comment by stefanosala** [comment]
Actually, that callback isn't being called in this case because it goes on [this line](https://github.com/fastify/deepmerge/blob/66ec946dac3448f83e15c70de5103a4cd7379aee/index.js#L134), where it is directly being merged into target.

In my example above, `target` is a plain object and `source` is an `EventEmitter`.

**Comment by segevfiner** [comment]
I used the logging-module as a workaround which is also annoying as I need to craft a separate file and compile it specially to share state with the main bundle and it also doesn't support ESM (#760)

I think the problem might be that `_deepmerge` in @fastify/deepmerge isn't checking for `cloneProtoObject` and the prototype before entering `mergeObject`, which then will recurse using `_deepmerge` into the logger property so it never actually reaches `cloneObject` or `cloneProtoObject`

**Comment by mcollina** [comment]
I guess the easiest way is for a volunteer to send a PR to fix.

**Comment by kilisamemarisaaa** [comment]
Verified on current `main` today while working through this issue — the situation has evolved since 2022:

**The deepmerge cloning described here is fixed.** `@fastify/deepmerge` v3 added a prototype check inside `mergeObject`, so with the `cloneProtoObject (obj) { return obj }` config `start.js` already passes, a Pino instance now survives the merge by reference:

```js
const deepmerge = require('@fastify/deepmerge')({ cloneProtoObject (obj) { return obj } })
const pino = require('pino')

const logger = pino({ transport: { target: 'pino-pretty' } })
const merged = deepmerge({ logger: { level: 'info' } }, { logger })

console.log(merged.logger === logger)      // true (the v2-era behaviour disassembled it)
console.log(typeof merged.logger.child)    // 'function'
```

**But the scenario still fails end-to-end on `main`, for a different reason**: fastify v5 rejects instances passed through the `logger` option (`FST_ERR_LOG_INVALID_LOGGER_CONFIG: logger options only accepts a configuration object`), and fastify-cli converges instances from `--logging-module` / `--options` files into `options.logger`. That part is addressed in #911 (route instances to `loggerInstance`), which builds on #910 — `fastify start` on main currently crashes earlier on unrelated ESM-interop regressions (`pkg-up@5` / `is-docker@4` / `chalk@6`) that #910 also fixes.

So in short: the original clone bug is gone (fixed by the deepmerge upgrade); the remaining instance-passing blocker is what #911 targets. Happy for maintainers to close as they see fit — thanks!
